core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -266,6 +275,158 @@ class KafkaController(val config: KafkaConfig,
+      var newFeatures: Features[FinalizedVersionRange] = Features.emptyFinalizedFeatures()
+      if (existingFeatureZNode.status.equals(FeatureZNodeStatus.Enabled)) {
+        newFeatures = Features.finalizedFeatures(existingFeatureZNode.features.features().asScala.map {
+          case (featureName, existingVersionRange) => {
+            val updatedVersionRange = defaultFinalizedFeatures.get(featureName)
+            if (updatedVersionRange == null) {
+              warn(s"Existing finalized feature: $featureName with $existingVersionRange"
+                + s" is absent in default finalized $defaultFinalizedFeatures")
+              (featureName, existingVersionRange)
+            } else if (existingVersionRange.max() >= updatedVersionRange.min()) {
+              // Through this change, we deprecate all version levels in the closed range:
+              // [existingVersionRange.min(), updatedVersionRange.min() - 1]
+              (featureName, new FinalizedVersionRange(updatedVersionRange.min(), existingVersionRange.max()))
+            } else {
+              // This is a special case: If the existing version levels fall completely outside the
+              // range of the default finalized version levels (i.e. no intersection), then, this
+              // case is not eligible for deprecation. This requires that the max version level be
+              // upgraded first to a value that's equal to the the default minimum version level.
+              info(s"Can not update minimum version level in finalized feature: $featureName,"
+              + s" since the existing $existingVersionRange does not intersect with the default"
+              + s" $updatedVersionRange.")
+              (featureName, existingVersionRange)
+            }
+          }
+        }.asJava)
+      }
+      val newFeatureZNode = new FeatureZNode(FeatureZNodeStatus.Enabled, newFeatures)
+      if (!newFeatureZNode.equals(existingFeatureZNode)) {
+        val newVersion = updateFeatureZNode(newFeatureZNode)
+        featureCache.waitUntilEpochOrThrow(newVersion, config.zkConnectionTimeoutMs)
+      }
+    }
+  }
+
+  /**
+   * Disables the feature versioning system (KIP-584).
+   *
+   * Sets up the FeatureZNode with disabled status. This status means the feature versioning system
+   * (KIP-584) is disabled, and, the finalized features stored in the FeatureZNode are not relevant.
+   * This status should be written by the controller to the FeatureZNode only when the broker
+   * IBP config is less than KAFKA_2_7_IV0.
+   *
+   * NOTE:
+   * 1. When this method returns, existing finalized features (if any) will be cleared from the
+   *    FeatureZNode.
+   * 2. This method, unlike enableFeatureVersioning() need not wait for the FinalizedFeatureCache
+   *    to be updated, because, such updates to the caceh (via FinalizedFeatureChangeListener)
+   *    are disabled when IBP config is < than KAFKA_2_7_IV0.
+   */
+  private def disableFeatureVersioning(): Unit = {
```

Comment:
Is it ok for us to always do `updateFeatureZNode`, since this call is idempotent?